crates/boojum/src/gadgets/tower_extension/algebraic_torus.rs

```diff
@@ -0,0 +1,494 @@
+    },
+};
+
+/// [`TorusWrapper`] is an algebraic compression of the `Fq12` element via underlying encoding of `Fq6`.
+/// In compressed form operations over Fq12 are less expensive.
+///
+/// The implementation is based on the following paper:
+/// https://eprint.iacr.org/2022/1162.pdf.
+#[derive(Clone, Debug, Copy)]
+pub struct TorusWrapper<F, T, NN, P>
+where
+    F: SmallField,
+    T: PrimeField,
+    NN: NonNativeField<F, T>,
+    P: TorusExtension12Params<T>,
+{
+    pub encoding: Fq6<F, T, NN, P::Ex6>,
+}
+
+// TODO: Probably, this could be implemented generally for any two Fqk and Fq(k/2) elements.
+impl<F, T, P, const N: usize> TorusWrapper<F, T, NonNativeFieldOverU16<F, T, N>, P>
+where
+    F: SmallField,
+    T: PrimeField,
+    P: TorusExtension12Params<T>,
+    [(); N + 1]:,
+{
+    /// Creates a new instance of the [`TorusWrapper`] with the given encoding.
+    pub fn new(encoding: Fq6<F, T, NonNativeFieldOverU16<F, T, N>, P::Ex6>) -> Self {
+        Self { encoding }
+    }
+
+    pub fn one<CS>(
+        cs: &mut CS,
+        params: &Arc<<NonNativeFieldOverU16<F, T, N> as NonNativeField<F, T>>::Params>,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let encoding = Fq6::zero(cs, params);
+        Self::new(encoding)
+    }
+
+    /// Returns the underlying parameters of the encoded `Fq6` element.
+    pub fn get_params(
+        &self,
+    ) -> &Arc<<NonNativeFieldOverU16<F, T, N> as NonNativeField<F, T>>::Params> {
+        self.encoding.get_params()
+    }
+
+    /// Normalizes the encoding of the `Fq6` element.
+    pub fn normalize<CS>(&mut self, cs: &mut CS)
+    where
+        CS: ConstraintSystem<F>,
+    {
+        self.encoding.normalize(cs);
+    }
+
+    /// Returns an instance if `flag` is `true`, otherwise returns a zero element.
+    pub fn mask<CS>(&mut self, cs: &mut CS, flag: Boolean<F>) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let zero = Fq6::zero(cs, self.get_params());
+        let new_encoding =
+            <Fq6<F, T, NonNativeFieldOverU16<F, T, N>, P::Ex6>>::conditionally_select(
+                cs,
+                flag,
+                &self.encoding,
+                &zero,
+            );
+
+        Self::new(new_encoding)
+    }
+
+    /// Compresses the `Fq12` element `c0 + c1*w` to the Torus (`T2`) element.
+    ///
+    /// Uses the formula `m <- (1 + c0) / c1` to compress the `Fq12` element with the additional
+    /// check for the exceptional case when `c1` is zero.
+    ///
+    /// If `SAFE=false`, then the function will not check for the exceptional case when `c1` is zero.
+    pub fn compress<CS, const SAFE: bool>(
+        cs: &mut CS,
+        f: &mut Fq12<F, T, NonNativeFieldOverU16<F, T, N>, P>,
+    ) -> Self
+    where
+        CS: ConstraintSystem<F>,
+    {
+        let params = f.get_params();
+        let mut c0 = f.c0.clone();
+        let mut c1 = f.c1.clone();
+
```

Comment:
Why do you use two versions here but didn't implement two versions for the other code?

Reply:
We based our torus implementation on `franklin-crypto`. There with flag `safe` are following methods:

1. `square`
2. `mul`
3. `pow`
4. `compress`

## `mul`

We did use what in `franklin-crypto` is marked as `unsafe` due to several reasons:

1. mathematically, as you mentioned, "unsafe" version is actually quite safe
2. computationally cheaper
3. easier to use and maintain from the engineering perspective

## `pow` 

This one has safe flag only because `square` has one. Below I'll explain why we didn't use it there, thus it is not added here.

## `square` and `compress` safeness:

These two actually have different taste of what unsafe flag is switching. While in square we potentially divide by zero, in `compress`, we at first make an enforcement that there won't be one.

## overall

In production settings, we use `safe` version.  I am struggling to imagine where the `unsafe` variant may be utilized, therefore we favoured the implementations of `safe` methods. 

The flag was added for `compress` because it is still quite `safe` compared to others, and we thought that it may be still usable in some production settings.

Therefore, if we desire to achieve some homogeneity of the codebase, we better to simply remove unsafe version from `compress` method, or, if this has some potential, leave it in there.